### ISSUE TITLE
find_package(Boost CONFIG REQUIRED) fails with CMake's CMP0057 related error if cmake_minimum_required is <= 3.2

### ISSUE THREAD

**Issue by traversaro** [conda-forge/boost-cpp-feedstock#103]
- [x] I read [the conda-forge documentation](https://conda-forge.org/docs/user/introduction.html#how-can-i-install-packages-from-conda-forge) and could not find the solution for my problem there.

Issue:

If one creates an environment with boost-cpp and the dev dependencies:
~~~
conda create -n boost-cpp boost-cpp compilers cmake pkg-config
~~~

And in this environment it configures a CMake project such as:
~~~
cmake_minimum_required(VERSION 3.2)

project(boost_mre)

find_package(Boost CONFIG REQUIRED)
~~~

The configuration fails with the error:
~~~
(boost-cpp) straversaro@IITICUBLAP102:~/boost_find_package_mre/build$ cmake ..
CMake Warning (dev) at /home/straversaro/mambaforge/envs/boost-cpp/lib/cmake/Boost-1.76.0/BoostConfig.cmake:240 (if):
  Policy CMP0057 is not set: Support new IN_LIST if() operator.  Run "cmake
  --help-policy CMP0057" for policy details.  Use the cmake_policy command to
  set the policy and suppress this warning.

  IN_LIST will be interpreted as an operator when the policy is set to NEW.
  Since the policy is not set the OLD behavior will be used.
Call Stack (most recent call first):
  CMakeLists.txt:6 (find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Error at /home/straversaro/mambaforge/envs/boost-cpp/lib/cmake/Boost-1.76.0/BoostConfig.cmake:240 (if):
  if given arguments:

    "ALL" "IN_LIST" "Boost_FIND_COMPONENTS"

  Unknown arguments specified
Call Stack (most recent call first):
  CMakeLists.txt:6 (find_package)


-- Configuring incomplete, errors occurred!
See also "/home/straversaro/boost_find_package_mre/build/CMakeFiles/CMakeOutput.log".
~~~

The problem is due to the fact that the `BoostConfig.cmake` file strongly assumes that the `IN_LIST` operator is available in CMake, so implicitly assuming that [`CMP0057`](https://cmake.org/cmake/help/latest/policy/CMP0057.html) is set to `ON`. This is true if `cmake_minimum_required` is at least 3.3, or if one calls `find_package(Boost REQUIRED)` without `CONFIG` or `NO_MODULE` options, as that will result in CMake's `FindBoost` being called, that sets CMP0057 to ON in https://github.com/Kitware/CMake/blob/v3.21.0/Modules/FindBoost.cmake#L594 .

However, if one has a project in which `cmake_minimum_required` is <= 3.2 and `find_package(Boost CONFIG)` or `find_package(Boost NO_MODULE)`  is invoked, it will fail with the error reported earlier in this issue. 

This can be fixed by patching (eventually even upstream) the  `BoostConfig.cmake` file to also enable explicitly CMP0057 as done in `FindBoost.cmake` module. However, in practice just bumping the `cmake_minimum_required` to 3.3 (CMake 3.3 was released in July 2015, ~ 6 years ago) or avoiding to use `find_package(Boost CONFIG)` will be sufficient to avoid this problem. For this reason, I do not plan to patch the feedstock, but I wanted to open an issue to clearly describe the problem if anyone was encountering it. However, if anyone is interested in patching Boost to fix the problem I would be happy to help.

<br/>
Environment (<code>conda list</code>):
<details>

```
$ conda list
(boost-cpp) straversaro@IITICUBLAP102:~/boost_find_package_mre/build$ conda list
# packages in environment at /home/straversaro/mambaforge/envs/boost-cpp:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                       1_gnu    conda-forge
binutils                  2.36.1               hdd6e379_1    conda-forge
binutils_impl_linux-64    2.36.1               h193b22a_1    conda-forge
binutils_linux-64         2.36                hf3e587d_32    conda-forge
boost-cpp                 1.76.0               h312852a_1    conda-forge
bzip2                     1.0.8                h7f98852_4    conda-forge
c-ares                    1.17.1               h7f98852_1    conda-forge
c-compiler                1.2.0                h7f98852_0    conda-forge
ca-certificates           2021.5.30            ha878542_0    conda-forge
cmake                     3.21.0               h8897547_0    conda-forge
compilers                 1.2.0                ha770c72_0    conda-forge
cxx-compiler              1.2.0                h4bd325d_0    conda-forge
expat                     2.4.1                h9c3ff4c_0    conda-forge
fortran-compiler          1.2.0                h1990efc_0    conda-forge
gcc_impl_linux-64         9.3.0               h70c0ae5_19    conda-forge
gcc_linux-64              9.3.0               hf25ea35_32    conda-forge
gfortran_impl_linux-64    9.3.0               hc4a2995_19    conda-forge
gfortran_linux-64         9.3.0               hdc58fab_32    conda-forge
gxx_impl_linux-64         9.3.0               hd87eabc_19    conda-forge
gxx_linux-64              9.3.0               h3fbe746_32    conda-forge
icu                       68.1                 h58526e2_0    conda-forge
kernel-headers_linux-64   2.6.32              h77966d4_13    conda-forge
krb5                      1.19.1               hcc1bbae_0    conda-forge
ld_impl_linux-64          2.36.1               hea4e1c9_1    conda-forge
libcurl                   7.77.0               h2574ce0_0    conda-forge
libedit                   3.1.20191231         he28a2e2_2    conda-forge
libev                     4.33                 h516909a_1    conda-forge
libgcc-devel_linux-64     9.3.0               h7864c58_19    conda-forge
libgcc-ng                 9.3.0               h2828fa1_19    conda-forge
libgfortran-ng            9.3.0               hff62375_19    conda-forge
libgfortran5              9.3.0               hff62375_19    conda-forge
libgomp                   9.3.0               h2828fa1_19    conda-forge
libnghttp2                1.43.0               h812cca2_0    conda-forge
libssh2                   1.9.0                ha56f1ee_6    conda-forge
libstdcxx-devel_linux-64  9.3.0               hb016644_19    conda-forge
libstdcxx-ng              9.3.0               h6de172a_19    conda-forge
libuv                     1.41.1               h7f98852_0    conda-forge
lz4-c                     1.9.3                h9c3ff4c_0    conda-forge
ncurses                   6.2                  h58526e2_4    conda-forge
openssl                   1.1.1k               h7f98852_0    conda-forge
pkg-config                0.29.2            h36c2ea0_1008    conda-forge
rhash                     1.4.1                h7f98852_0    conda-forge
sysroot_linux-64          2.12                h77966d4_13    conda-forge
tk                        8.6.10               h21135ba_1    conda-forge
xz                        5.2.5                h516909a_1    conda-forge
zlib                      1.2.11            h516909a_1010    conda-forge
zstd                      1.5.0                ha95c52a_0    conda-forge
```
</details>

<br/>
Details about  <code>conda</code> and system ( <code>conda info</code> ):
<details>

```
$ conda info
(boost-cpp) straversaro@IITICUBLAP102:~/boost_find_package_mre/build$ conda info

     active environment : boost-cpp
    active env location : /home/straversaro/mambaforge/envs/boost-cpp
            shell level : 1
       user config file : /home/straversaro/.condarc
 populated config files : /home/straversaro/mambaforge/.condarc
                          /home/straversaro/.condarc
          conda version : 4.10.3
    conda-build version : not installed
         python version : 3.9.5.final.0
       virtual packages : __linux=4.19.104=0
                          __glibc=2.31=0
                          __unix=0=0
                          __archspec=1=x86_64
       base environment : /home/straversaro/mambaforge  (writable)
      conda av data dir : /home/straversaro/mambaforge/etc/conda
  conda av metadata url : None
           channel URLs : https://conda.anaconda.org/conda-forge/linux-64
                          https://conda.anaconda.org/conda-forge/noarch
          package cache : /home/straversaro/mambaforge/pkgs
                          /home/straversaro/.conda/pkgs
       envs directories : /home/straversaro/mambaforge/envs
                          /home/straversaro/.conda/envs
               platform : linux-64
             user-agent : conda/4.10.3 requests/2.25.1 CPython/3.9.5 Linux/4.19.104-microsoft-standard ubuntu/20.04.2 glibc/2.31
                UID:GID : 1000:1000
             netrc file : None
           offline mode : False
```
</details>

**Comment by traversaro** [comment]
fyi @kunaltyagi @tobias-Fischer

**Comment by traversaro** [comment]
This issue was also creating the problem documented in this Stack Overflow question: https://stackoverflow.com/questions/62930429/c-avro-cmake-failed .

**Comment by traversaro** [comment]
> For this reason, I do not plan to patch the feedstock, but I wanted to open an issue to clearly describe the problem if anyone was encountering it. However, if anyone is interested in patching Boost to fix the problem I would be happy to help.

If anyone is interested in this, I proposed a fix upstream in https://github.com/boostorg/boost_install/pull/52 .

**Comment by jakirkham** [comment]
It looks like that fix went into Boost 1.78.0, which we just merged in PR ( https://github.com/conda-forge/boost-cpp-feedstock/pull/121 ) and should publish packages for soon. Is there anything else to do here @traversaro?

**Comment by traversaro** [comment]
No, everything should be fixed, thanks!
